nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/ValidateJson.java

```diff
@@ -0,0 +1,215 @@
+        FlowFile flowFile = session.get();
+        if ( flowFile == null ) {
+            return;
+        }
+
+        final AtomicReference<Exception> exceptions = new AtomicReference<>();
+        final AtomicReference<Set<ValidationMessage>> validationErrors = new AtomicReference<Set<ValidationMessage>>(null);
+
+        session.read(flowFile, new InputStreamCallback()  {
+
+            @Override
+            public void process(InputStream in) {
+                try {
+                    // Set JSON schema version to use from processor property
+                    VersionFlag schemaVersion = VersionFlag.V201909;
+                    if (context.getProperty(SCHEMA_VERSION).getValue() == SCHEMA_VERSION_4.getValue()) {
+                        schemaVersion = VersionFlag.V4;
+                    }
+                    if (context.getProperty(SCHEMA_VERSION).getValue() == SCHEMA_VERSION_6.getValue()) {
+                        schemaVersion = VersionFlag.V6;
+                    }
+                    if (context.getProperty(SCHEMA_VERSION).getValue() == SCHEMA_VERSION_7.getValue()) {
+                        schemaVersion = VersionFlag.V7;
+                    }
+                    if (context.getProperty(SCHEMA_VERSION).getValue() == SCHEMA_VERSION_V201909.getValue()) {
+                        schemaVersion = VersionFlag.V201909;
+                    }
+        
+                    // Read in flowFile inputstream, and validate against schema
+                    ObjectMapper mapper = new ObjectMapper();
+                    JsonNode node = mapper.readTree(in);
+                    String schemaText = context.getProperty(SCHEMA_TEXT).evaluateAttributeExpressions().getValue();
+                    JsonSchemaFactory factory = JsonSchemaFactory.getInstance(schemaVersion);
+                    JsonSchema schema = factory.getSchema(schemaText);
+                    validationErrors.set(schema.validate(node));
+                
+                } catch (JsonParseException jpe) {
+                    exceptions.set(jpe);
+                } catch (IOException ioe) {
+                    exceptions.set(ioe);
+                }
+            }
+        });
+
+        // Failed to read flowFile - either IOException, or JsonParseException
+        if (exceptions.get() != null) {
+            this.getLogger().info("Failed to process {} due to {}; routing to 'failure'", new Object[]{flowFile, exceptions.get().getLocalizedMessage()});
+            session.getProvenanceReporter().route(flowFile, REL_FAILURE);
+            session.transfer(flowFile, REL_FAILURE);
+
+        // Schema checks failed
+        } else if (validationErrors.get().size() > 0) {
+            flowFile = session.putAttribute(flowFile, ERROR_ATTRIBUTE_KEY, validationErrors.get().toString());
+            this.getLogger().info("Failed to validate {} against schema; routing to 'invalid'", new Object[]{flowFile});
```

Comment:
This is certainly a bit nit-picky i realize, but indicating "Failed to validate" sounds like there was an error or something when attempting to perform validation. Validation went fine, the data just happened not to be valid. May want to change to something a bit clearer like "Found {} to be invalid when validated against schema; routing to 'invalid'"?

Reply:
Not nit-picky! More than happy to change to your wording